src/errors.rs

```diff
@@ -0,0 +1,180 @@
+//! Error definitions
+
+pub mod i2c {
+    /// I2C error
+    pub trait Error: core::fmt::Debug {
+        /// Convert error to a generic I2C error kind
+        ///
+        /// By using this method I2C errors freely defined by HAL implementations
+        /// can be converted to a set of generic I2C errors upon which generic
+        /// code can act.
+        fn kind(&self) -> ErrorKind;
+    }
+
+    /// I2C error kind
+    ///
+    /// This represents a common set of I2C operation errors. HAL implementations are
+    /// free to define more specific or additional error types. However, by providing
+    /// a mapping to these common I2C errors, generic code can still react to them.
+    #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+    #[non_exhaustive]
+    pub enum ErrorKind {
+        /// An unspecific bus error occurred
+        Bus,
+        /// The arbitration was lost, e.g. electrical problems with the clock signal
+        ArbitrationLoss,
+        /// A bus operation was not acknowledged, e.g. due to the addressed device not being available on
+        /// the bus or the device not being ready to process requests at the moment
+        NoAcknowledge,
+        /// The peripheral receive buffer was overrun
+        Overrun,
+        /// The peripheral send buffer ran out of data
+        Underrun,
+        /// A different error occurred. The original error may contain more information.
+        Other,
+    }
+
+    impl Error for ErrorKind {
+        fn kind(&self) -> ErrorKind {
+            *self
+        }
+    }
```

Comment:
This is elegant 👍🏻 Quick-and-dirty HAL implementations can use the `ErrorKind` directly as error, while a more feature-rich HAL would implement a custom error type with appropriate conversions.